File: Feb-2023/Java/MinimizeDeviationInArray.java
// 1675. Minimize Deviation in Array
class Solution {
    public int minimumDeviation(int[] nums) {
        TreeSet<Integer> set = new TreeSet<>();
       // Take care of all odd conditions by multiplying all odd to even
        for(int i=0; i<nums.length; i++){
            if(nums[i]%2==1) nums[i]*=2;
            set.add(nums[i]); 
        }
        //treeSet to get the elements in logn average complexity
        //Elements in treeset are unique and sorted
        int diff=Integer.MAX_VALUE;
        while(true){
            //First element is the smallest because it is sorted, get the maximum value by using set.last() method
            int small = set.first();
            int large = set.last();
            diff = Math.min(diff,large-small);
            //The set still can an even number which is the largest that can be divided 
            if(large%2==0){
                set.remove(large);
                set.add(large/2);
            }
            else{
                return diff;
            }
            
        }
      
    }
}

File: Feb-2023/C++/MinimizeDeviationInArray.cxx

// 1675. Minimize Deviation in Array
class Solution {
public:
    int minimumDeviation(vector<int>& nums) {
        set<int> set;
        // Take care of all odd conditions by multiplying all odd to even
        for(int i = 0; i < nums.size(); i++) {
            if(nums[i] % 2 == 1) nums[i] *= 2;
            set.insert(nums[i]);
        }
        // set to get the elements in logn average complexity
        // Elements in set are unique and sorted
        int diff = INT_MAX;
        while(true) {
            // First element is the smallest because it is sorted, get the maximum value by using set.rbegin() method
            int small = *set.begin();
            int large = *set.rbegin();
            diff = min(diff, large - small);
            // The set still has an even number which is the largest that can be divided
            if(large % 2 == 0) {
                set.erase(large);
                set.insert(large / 2);
            } else {
                return diff;
            }
        }
    }
};
